daemon/container_config.c

```diff
@@ -148,33 +154,139 @@ container_config_storage_size(container_config_t *config)
+				sig = NULL;
+			}
+		}
+		cert_size = file_size(cert_file);
+		if (cert_size > 0) {
+			cert = mem_alloc(cert_size);
+			if (-1 == file_read(cert_file, (char *)cert, cert_size)) {
+				ERROR("Failed to read cert file '%s'!", cert_file);
+				mem_free(cert);
+				cert = NULL;
+			}
+		}
+		mem_free(sig_file);
+		mem_free(cert_file);
+	}
+
+	// check cert and signature buffers
+	IF_TRUE_GOTO(cert_size <= 0 || sig_size <= 0 || cert == NULL || sig == NULL, out);
+
+	smartcard_crypto_verify_result_t verify_result = smartcard_crypto_verify_buf_block(
+		conf_buf, conf_len, sig, sig_size, cert, cert_size, C_CONFIG_VERIFY_HASH_ALGO);
+
+	ret = (verify_result == VERIFY_GOOD) ? true : false;
+out:
+	INFO("Verify Result of target with prefix '%s': %s", prefix, ret ? "GOOD" : "UNSIGNED");
+
+	mem_free(sig);
+	mem_free(cert);
+	return ret;
+}
+
 container_config_t *
-container_config_new(const char *file, const uint8_t *buf, size_t len)
+container_config_new(const char *file, const uint8_t *buf, size_t len, uint8_t *sig_buf,
+		     size_t sig_len, uint8_t *cert_buf, size_t cert_len)
 {
 	ContainerConfig *ccfg = NULL;
+	uint8_t *buf_internal = NULL;
+	container_config_t *config = NULL;
 
 	ASSERT(file);
+	off_t conf_len = len;
 
-	if (buf) {
+	char *prefix = mem_strdup(file);
+	size_t file_len = strlen(file);
+	prefix[file_len - 5] = '\0';
+	IF_TRUE_GOTO(file_len < 5 || strcmp(file + file_len - 5, ".conf"), out);
+
+	// check if config comes from buffer or needs to be read from file
+	if (buf == NULL) {
+		DEBUG("Loading container config from file \"%s\".", file);
+		conf_len = file_size(file);
+		if (conf_len > 0) {
+			buf_internal = mem_alloc(conf_len);
+			if (-1 == file_read(file, (char *)buf_internal, conf_len)) {
+				mem_free(buf_internal);
+				buf_internal = NULL;
+			}
+		}
+	} else {
 		DEBUG("Loading container config from buf storing to file \"%s\".", file);
-		ccfg = (ContainerConfig *)protobuf_message_new_from_buf(
-			buf, len, &container_config__descriptor);
-		if (!ccfg)
-			WARN("Failed loading container config from buf");
+		buf_internal = mem_new0(uint8_t, conf_len);
+		memcpy(buf_internal, buf, conf_len);
+	}
+
+	if (!container_config_verify(prefix, buf_internal, conf_len, sig_buf, sig_len, cert_buf,
+				     cert_len)) {
+		ERROR("Failed verify signature of container config for file \"%s\".", file);
+		goto out;
 	}
+
+	ccfg = (ContainerConfig *)protobuf_message_new_from_buf(buf_internal, conf_len,
+								&container_config__descriptor);
 	if (!ccfg) {
-		DEBUG("Loading container config from file \"%s\".", file);
-		ccfg = (ContainerConfig *)protobuf_message_new_from_textfile(
-			file, &container_config__descriptor);
-		if (!ccfg) {
-			WARN("Failed loading container config from file \"%s\".", file);
-			return NULL;
+		WARN("Failed loading container config from buf");
+		goto out;
+	}
+
+	// if config was provided by buf, update all files according to buffers
+	if (buf) {
+		if (-1 == file_write(file, (char *)buf, conf_len))
```

Comment:
nop because we check if we got here from config buffer if a call to create_config was made with a config in from of file we do need to do anything, since this is the old behavior.